vj4/model/domain.py

```diff
@@ -106,6 +106,26 @@ def get_pending(**kwargs):
                                         return_document=ReturnDocument.AFTER)
 
 
+@argmethod.wrap
+async def inc_pid_counter(domain_id):
+  """Increments the problem ID counter.
+
+  Returns:
+    Integer value after increment.
+  """
+  coll = db.coll('domain')
+  doc = await coll.find_one_and_update(filter={'_id': domain_id},
+                                       update={'$inc': {'pid_counter': 1}},
+                                       upsert=True,
+                                       return_document=ReturnDocument.AFTER)
+  if doc['pid_counter'] == 1:
```

Comment:
Bug if somehow 2 or more requests reach this point concurrently.

Reply:
You can use `$setOnInsert` to set the default value. See https://docs.mongodb.com/manual/reference/operator/update/setOnInsert/ .

However, a unique index is needed for atomicity. Also see discussion at https://grokbase.com/t/gg/mongodb-user/14aeaptjb7/findandupdate-atomicity-without-a-unique-index-im-stuck .

Reply:
Actually `upsert` and the unique index is not needed because domain doc must exist and `$setOnInsert` should not work.

Reply:
I think the solution here should work but it's crowded
https://stackoverflow.com/questions/29746158/mongodb-counter-with-initial-value-not-zero

Other alternatives:
* add 1000 to the return value
* start from 1
* update the database offline

Reply:
Use "set if not exists" finally